File: Source/View/AKTOutlineView.m
//----------------------------------------------------------------------------------------------------------------------
//	AKTOutlineView.m			©2026 Stevo Brock	All rights reserved.
//----------------------------------------------------------------------------------------------------------------------

#import "AKTOutlineView.h"

//----------------------------------------------------------------------------------------------------------------------
// MARK: AKTOutlineView

@implementation AKTOutlineView

// MARK: NSResponder methods

//----------------------------------------------------------------------------------------------------------------------
- (void) mouseDown:(NSEvent*) event
{
	// Setup
	NSPoint			point = [self convertPoint:event.locationInWindow fromView:nil];
	NSInteger		rowIndex = [self rowAtPoint:point];
	NSInteger		columnIndex = [self columnAtPoint:point];
	NSTableColumn*	tableColumn =
							((columnIndex >= 0) && (columnIndex < (NSInteger) self.tableColumns.count)) ?
									self.tableColumns[columnIndex] : nil;
	id				item = [self itemAtRow:rowIndex];

	if (self.isOptionClickToBeginEditingEnabled &&
			(event.modifierFlags & NSEventModifierFlagDeviceIndependentFlagsMask) == NSEventModifierFlagOption) {
		// Start editing on option-click
		if ([self.delegate outlineView:self shouldEditTableColumn:tableColumn item:item])
			[self editColumn:columnIndex row:rowIndex withEvent:event select:YES];
	} else if (self.isDoubleClickToBeginEditingEnabled && event.clickCount == 2) {
		// Start editing on double-click if allowed, otherwise fall through to default
		if ([self.delegate outlineView:self shouldEditTableColumn:tableColumn item:item])
			[self editColumn:columnIndex row:rowIndex withEvent:event select:YES];
		else
			[super mouseDown:event];
	} else
		[super mouseDown:event];
}

@end

File: Source/View/AKTOutlineView.swift
//
//  AKTOutlineView.swift
//  AppKit Toolbox
//
//  Created by Stevo Brock on 11/2/24.
//  Copyright © 2024 Stevo Brock. All rights reserved.
//

import AppKit

//----------------------------------------------------------------------------------------------------------------------
// MARK: AKTOutlineView
public class AKTOutlineView : NSOutlineView {

	// MARK: Types
	public typealias MenuProc = (_ row :Int, _ tableColumnIndex :Int, _ tableColumn :NSTableColumn?) -> NSMenu?

	// MARK: Properties
	public	var	menuProc :MenuProc? = nil

	// MARK: NSView methods
	//------------------------------------------------------------------------------------------------------------------
	override public func menu(for event :NSEvent) -> NSMenu? {
		// Check if have menuProc
		if let menuProc = self.menuProc {
			// Setup
			let	point = convert(event.locationInWindow, from: nil)
			let	columnIndex = column(at: point)

			return menuProc(row(at: point), columnIndex, 
					((columnIndex >= 0) && (columnIndex < self.tableColumns.count)) ?
							self.tableColumns[columnIndex] : nil)
		} else {
			// No menuProc
			return self.menu
		}
	}
}
